File: Math/223_RectangleArea.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com


class Solution(object):
    def computeArea(self, A, B, C, D, E, F, G, H):
        size_1 = (C-A) * (D-B)
        size_2 = (G-E) * (H-F)
        left = max(A, E)
        bottom = max(B, F)
        right = min(C, G)
        top = min(D, H)

        # There is an area coverd by both the two rectangle
        if left < right and bottom < top:
            return size_1 + size_2 - (top-bottom) * (right-left)
        else:
            return size_1 + size_2

"""
-2
-2
2
2
-2
-2
2
2
0
0
0
0
-1
-1
1
1
"""

File: Math/223_RectangleArea.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-12 09:17:10
 */

class Solution {
public:
    int computeArea(int A, int B, int C, int D, int E, int F, int G, int H) {
        int left = max(A, E);
        int bottom = max(B, F);
        int right = min(C, G);
        int top = min(D, H);
        int size = (D-B) * (C-A) + (H-F) * (G-E);
        if(left < right && bottom < top){
            size -= (top-bottom) * (right-left);
        }
        return size;
    }
};

/*
-2
-2
2
2
-2
-2
2
2
0
0
0
0
-1
-1
1
1
*/
